### PR TITLE
Buffer

### PR DESCRIPTION
**Pull request by saisunku**
## Contributor Checklist:

* [x] I have added myself to [the creators.json file](https://github.com/LiberTEM/LiberTEM/blob/master/packaging/creators.json)
* [X] I have added [a changelog entry](https://github.com/LiberTEM/LiberTEM/tree/master/docs/source/changelog) for my contribution
* [X] I have added/updated documentation for all user-facing changes
* [X] I have added/updated test cases

### CODE REVIEW

File: src/libertem/common/buffers.py [hunk 1]
```diff
@@ -62,17 +63,24 @@ def __init__(self, kind, extra_shape=(), dtype="float32"):
...
+            For a Shape object, sig_dims is discarded and the entire shape is used.
 
         dtype : string or numpy dtype
             The dtype of this buffer
         """
+
+        if isinstance(extra_shape, Shape):
+            self._extra_shape = extra_shape._nav_shape + extra_shape._sig_shape
+        else:
+            self._extra_shape = extra_shape
```
Comment (sk1p) [inline src/libertem/common/buffers.py]: Even though it works, accessing internal attributes of other classes (starting with an underscore) should not be done - they are used to separate the interface from implementation details (in other languages, there are stricter mechanisms for ensuring this, like visibility specifiers `private` etc.). Imagine the `Shape` implementation is changed to store the whole shape as a single tuple (or even another type: list, ndarray?) - you would have to change `_nav_shape` and `_sig_shape` usages all over the code base.

There is an even better way to do the same: `Shape` implements `__iter__`, and  any iterable can be converted to a `tuple` using the `tuple()` built-in function. So there is also no need for the `isinstance` check anymore - just unconditionally call `tuple` on `extra_shape`.

That means we accept any iterable, not just tuple and `Shape`, [which is even better IMHO.](https://en.wikipedia.org/wiki/Robustness_principle)

File: docs/source/changelog/bugfix/buffer_extra_shape.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+[Bugfix] Allow BufferWrapper to accept a Shape object
+===========================================================
+
+* Both tuples and Shape objects can be used as extra_shape now
```
[the remark this answers is not in the document]
Comment (saisunku) [reply docs/source/changelog/bugfix/buffer_extra_shape.rst]: Done

File: src/libertem/common/buffers.py (hunk 1 above)
Comment (saisunku) [reply src/libertem/common/buffers.py]: Thanks for the detailed explanation, I've updated accordingly

**Comment by sk1p** [comment]
Hm, for some reason the travis build doesn't show up in the PR, sorry for that. This is the associated travis build:

https://travis-ci.org/github/LiberTEM/LiberTEM/builds/668279968

Note that [the flake8 job](https://travis-ci.org/github/LiberTEM/LiberTEM/jobs/668279974) failed - there are a few style fixes to be done. Please see our [contributing guideline](https://libertem.github.io/LiberTEM/contributing.html#code-style) on how to run them locally.

The mypy job that also failed is allowed to fail - don't worry about that one. Other than that, it looks ready to merge!

**Comment by saisunku** [comment]
Thanks! I had some issues with `tox`, so I ran `pytest` manually which didn't include `flake8`. Fixed now.

**Comment by sk1p** [comment]
Thanks, merged!

**Pull request merged**